video/out/vo_drm.c

```diff
@@ -0,0 +1,509 @@
+
+    struct mp_rect src_rc = p->src;
+    src_rc.x0 = MP_ALIGN_DOWN(src_rc.x0, mpi->fmt.align_x);
+    src_rc.y0 = MP_ALIGN_DOWN(src_rc.y0, mpi->fmt.align_y);
+    mp_image_crop_rc(mpi, src_rc);
+
+    mp_sws_scale(p->sws, p->curframe, mpi);
+
+    osd_draw_on_image(vo->osd, p->osd, mpi ? mpi->pts : 0, 0, p->curframe);
+
+    int32_t shift = (p->device_w * p->y + p->x) * 4;
+    memcpy_pic(front_buf->map + shift,
+               p->curframe->planes[0],
+               (p->dst.x1 - p->dst.x0) * 4,
+               p->device_h,
+               p->device_w * 4,
+               p->curframe->stride[0]);
+
+    talloc_free(mpi);
+}
+
+static void flip_page(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+
+    int ret = drmModeSetCrtc(p->fd, p->dev->crtc,
+                             p->dev->bufs[p->dev->front_buf].fb,
+                             0, 0, &p->dev->conn, 1, &p->dev->mode);
+    if (ret) {
+        MP_WARN(vo, "Cannot flip page for connector\n");
+    } else {
+        p->dev->front_buf ++;
+        p->dev->front_buf %= BUF_COUNT;
+    }
+}
+
+static int preinit(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+    p->dev = NULL;
+    p->fd = 0;
+    p->old_crtc = NULL;
+    p->curframe = NULL;
+    p->sws = mp_sws_alloc(vo);
+
+    int ret;
+
+    ret = modeset_open(vo, &p->fd, p->device_path);
+    if (ret)
+        return ret;
+
+    ret = modeset_prepare_dev(vo, p->fd, p->connector_id, &p->dev);
+    if (ret)
+        return ret;
+
+    assert(p->dev);
+    p->device_w = p->dev->bufs[0].width;
+    p->device_h = p->dev->bufs[0].height;
+
+    p->old_crtc = drmModeGetCrtc(p->fd, p->dev->crtc);
+    ret = drmModeSetCrtc(p->fd, p->dev->crtc,
+                         p->dev->bufs[p->dev->front_buf + BUF_COUNT - 1].fb,
+                         0, 0, &p->dev->conn, 1, &p->dev->mode);
+    if (ret) {
+        MP_ERR(vo, "Cannot set CRTC for connector %u: %s\n", p->connector_id,
+               mp_strerror(errno));
+    }
+
+    return 0;
+}
+
+static void uninit(struct vo *vo)
+{
+    struct priv *p = vo->priv;
+
+    if (p->dev) {
+        if (p->old_crtc) {
+            drmModeSetCrtc(p->fd,
+                           p->old_crtc->crtc_id,
+                           p->old_crtc->buffer_id,
+                           p->old_crtc->x,
+                           p->old_crtc->y,
+                           &p->dev->conn,
+                           1,
+                           &p->dev->mode);
+            drmModeFreeCrtc(p->old_crtc);
+        }
+
+        modeset_destroy_fb(p->fd, &p->dev->bufs[1]);
+        modeset_destroy_fb(p->fd, &p->dev->bufs[0]);
+    }
+
+    talloc_free(p->curframe);
+    talloc_free(p->dev);
+}
+
+static int query_format(struct vo *vo, int format)
+{
+    if (sws_isSupportedInput(imgfmt2pixfmt(format)))
+        return 1;
+    return format == IMGFMT_BGR0;
+}
+
+static int control(struct vo *vo, uint32_t request, void *data)
+{
+    return VO_NOTIMPL;
+}
```

Comment:
Did you test whether redrawing works? VOCTRL_REDRAW_FRAME is not implemented, but it should be emulated by sending the full frame again. However, I'm not sure if it really works, and it won't work if there's no video frame yet. (If you use ``--force-window --idle``, the OSD is useable even though no video frame was sent; although this might be very useless in this mode, so maybe it doesn't matter. But the same situation can happen if you e.g. seek, and network slowness or whatever prevents that a new video frame can be sent soon.)